DEFI/LiquidityPool/certora/specs/Full.spec

```diff
@@ -0,0 +1,448 @@
+
+    uint256 UnderlyingBalanceBefore = underlying.balanceOf(e.msg.sender);
+    uint256 SharesBefore = balanceOf(e.msg.sender);
+    
+    calldataarg args;
+    f(e, args);
+    
+    uint256 UnderlyingBalanceAfter = underlying.balanceOf(e.msg.sender);
+    uint256 SharesAfter = balanceOf(e.msg.sender);
+    
+    assert UnderlyingBalanceBefore < UnderlyingBalanceAfter <=> SharesBefore > SharesAfter;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ More shares leads to bigger withdraw                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule moreSharesMoreWithdraw(env e) {
+    uint256 sharesX;
+    uint256 sharesY;
+    uint256 amountX;
+    uint256 amountY;
+
+    storage init = lastStorage;
+    
+    amountX = withdraw(e, sharesX);
+    amountY = withdraw(e, sharesY) at init;
+    
+    assert sharesX > sharesY => amountX >= amountY;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rounding favours the house                                                                                          │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule amountRoundingTripFavoursContract(env e) {
+    requireInvariant totalSharesIsZeroWithUnderlying();
+
+    uint256 clientAmountBefore = underlying.balanceOf(e.msg.sender);
+    uint256 contractAmountBefore = underlying.balanceOf(currentContract);
+
+    uint256 clientShares = deposit(e, clientAmountBefore);
+    uint256 clientAmountAfter = withdraw(e, clientShares);
+    uint256 contractAmountAfter = underlying.balanceOf(currentContract);
+
+    assert clientAmountBefore >= clientAmountAfter;
+    assert contractAmountBefore <= contractAmountAfter;
+}
+
+/*
+Bug detected, if contract deploy with balance it can be drained. 
+*/
+rule sharesRoundingTripFavoursContract(env e) {
+    uint256 clientSharesBefore = balanceOf(e.msg.sender);
+    uint256 contractSharesBefore = balanceOf(currentContract);
+
+    requireInvariant totalSharesEqualSumOfShares();
+    require underlying.balanceOf(currentContract) > 0 => balanceOf(currentContract) > 0; //this assumption hides the bug and cant be proved.
+    require sumOfShares >= clientSharesBefore + contractSharesBefore;
+    require e.msg.sender != currentContract; // this assumption must hold to avoid shares dilute attack
+
+
+    uint256 clientAmount = withdraw(e, clientSharesBefore);
+    uint256 clientSharesAfter = deposit(e, clientAmount);
+    uint256 contractSharesAfter = balanceOf(currentContract);
+
+    assert clientSharesBefore >= clientSharesAfter;
+    assert contractSharesBefore <= contractSharesAfter;
+}
```

Comment:
lets fix the code that this rule passes, we can have a broken version and fixed version

Reply:
How do you suggest fixing it because the only way I see is to scam the first depositor :)

Reply:
I will try a patch of remember the last withdraw ratio and use it for next depositor

Reply:
Tried  deposited amont patch